core/kernel/lockdep.c

```diff
@@ -0,0 +1,397 @@
+				   struct lockdep_node *to,
+				   vaddr_t *call_stack_from,
+				   vaddr_t *call_stack_to,
+				   uintptr_t thread_id)
+{
+	struct lockdep_edge *edge = NULL;
+
+	STAILQ_FOREACH(edge, &from->edges, link)
+		if (edge->to == to)
+			return TEE_SUCCESS;
+
+	edge = calloc(1, sizeof(*edge));
+	if (!edge)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	edge->to = to;
+	edge->call_stack_from = dup_call_stack(call_stack_from);
+	edge->call_stack_to = dup_call_stack(call_stack_to);
+	edge->thread_id = thread_id;
+	STAILQ_INSERT_TAIL(&from->edges, edge, link);
+
+	return TEE_SUCCESS;
+}
+
+struct lockdep_bfs {
+	struct lockdep_node *node;
+	uintptr_t *path;
+	int pathlen;
+	TAILQ_ENTRY(lockdep_bfs) link;
+};
+
+TAILQ_HEAD(lockdep_bfs_head, lockdep_bfs);
+
+static void lockdep_bfs_queue_delete(struct lockdep_bfs_head *queue)
+{
+	struct lockdep_bfs *cur = NULL;
+	struct lockdep_bfs *next = NULL;
+
+	TAILQ_FOREACH_SAFE(cur, queue, link, next) {
+		TAILQ_REMOVE(queue, cur, link);
+		free(cur->path);
+		free(cur);
+	}
+}
+
+/*
+ * Print shortest cycle in @graph that contains @node.
+ * This function performs an iterative breadth-first search starting from @node,
+ * and stops when it reaches @node again. In each node we're tracking the path
+ * from the start node.
+ */
+static uintptr_t *lockdep_graph_get_shortest_cycle(struct lockdep_node *node)
+{
+	struct lockdep_bfs_head queue;
+	struct lockdep_bfs *qe = NULL;
+	uintptr_t *ret = NULL;
+
+	TAILQ_INIT(&queue);
+	node->flags |= LOCKDEP_NODE_BFS_VISITED;
+
+	qe = calloc(1, sizeof(*qe));
+	if (!qe)
+		goto out;
+	qe->node = node;
+	qe->path = malloc(sizeof(uintptr_t));
+	if (!qe->path)
+		goto out;
+	qe->path[0] = node->lock_id;
+	qe->pathlen = 1;
+	TAILQ_INSERT_TAIL(&queue, qe, link);
+
+	while (!TAILQ_EMPTY(&queue)) {
+		qe = TAILQ_FIRST(&queue);
+
+		struct lockdep_node *n = qe->node;
+
+		TAILQ_REMOVE(&queue, qe, link);
+
+		struct lockdep_edge *e = NULL;
+
+		STAILQ_FOREACH(e, &n->edges, link) {
+			if (e->to->lock_id == node->lock_id) {
+				uintptr_t *tmp = NULL;
+				size_t nlen = qe->pathlen + 1;
+
+				/*
+				 * Cycle found. Terminate cycle path with NULL
+				 * and return it.
+				 */
+				tmp = realloc(qe->path,
+					      nlen * sizeof(uintptr_t));
+				if (!tmp)
+					goto out;
+				qe->path = tmp;
+				qe->path[nlen - 1] = 0;
+				ret = qe->path;
+				goto out;
+			}
+
+			if (!(e->to->flags & LOCKDEP_NODE_BFS_VISITED)) {
+				e->to->flags |= LOCKDEP_NODE_BFS_VISITED;
+
+				size_t nlen = qe->pathlen + 1;
+				struct lockdep_bfs *nqe = calloc(1,
+								 sizeof(*nqe));
+				if (!nqe)
+					goto out;
+				nqe->node = e->to;
+				nqe->path = malloc(nlen * sizeof(uintptr_t));
+				if (!nqe->path)
+					goto out;
+				nqe->pathlen = nlen;
+				memcpy(nqe->path, qe->path,
+				       qe->pathlen * sizeof(uintptr_t));
+				nqe->path[nlen - 1] = e->to->lock_id;
+				TAILQ_INSERT_TAIL(&queue, nqe, link);
+			}
+		}
+		free(qe->path);
+		free(qe);
+		qe = NULL;
+	}
+
+out:
+	free(qe);
```

Comment:
`free(qe->path);` also, we have several `goto out` in other places after a successful malloc.

Reply:
Ah, wait a bit `ret = qe->path;` ... kind of hides that. So my previous comment here might be an incorrect claim.

Reply:
If allocation fails at line 157, the caller may print out incomplete information without noticing the error.
Should this function return NULL in this case and locally free the allocated `qe->path`?